File: Data Structures & Algorithms/trapping-rain-water/submission-1.cs
public class Solution {
    // Standard: intuitive, two max trackers
public int Trap(int[] height) {
    int left = 0, right = height.Length - 1;
    int leftMax = 0, rightMax = 0, water = 0;
    while (left < right) {
        if (height[left] < height[right]) {
            leftMax = Math.Max(leftMax, height[left]);
            water += leftMax - height[left++];
        } else {
            rightMax = Math.Max(rightMax, height[right]);
            water += rightMax - height[right--];
        }
    }
    return water;
}
}

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int maxSum = -1e9;

    int maxPathSumHelper(TreeNode* root)
    {
        if(!root) return 0;

        int leftMax = maxPathSumHelper(root->left);
        int rightMax = maxPathSumHelper(root->right);

        if(leftMax < 0) leftMax = 0;
        if(rightMax < 0) rightMax = 0;

        maxSum = max(maxSum, root->val+leftMax+rightMax);

        return root->val+max(leftMax, rightMax);
    }

    int maxPathSum(TreeNode* root) {
        
        maxPathSumHelper(root);

        return maxSum;
    }
};
